### PR TITLE
[FLINK-19779][avro] Remove the record_ field name prefix for Confluen…

### PR DESCRIPTION
**Pull request by danny0405**
…t Avro format deserialization

## What is the purpose of the change

Add prefix for the field name only when the record and field have the same name.

## Brief change log

  - Modify `AvroSchemaConverter.convertToSchema` to only append field prefix when the record and field have the same name
  - Modify the existing test

## Verifying this change

Added UT.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no

### CODE REVIEW

File: flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroSchemaConverter.java
Comment (V1ncentzzZ) [inline flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroSchemaConverter.java:376]: I think there is still existed a bug. see: https://issues.apache.org/jira/browse/FLINK-19779?focusedCommentId=17219588&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17219588

**Comment by wuchong** [comment]
Let's hold this PR a bit and continue discussion in FLINK-19779.

File: flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroSchemaConverter.java
Comment (danny0405) [reply flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroSchemaConverter.java:376]: Yes, the avro schema builder does not allow same name field names, even if they are in different scope (different layer).

Comment (dawidwys) [reply flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroSchemaConverter.java:383]: I think the correct solution will be:

```
				RowType rowType = (RowType) logicalType;
				List<String> fieldNames = rowType.getFieldNames();
				// we have to make sure the record name is different in a Schema
				SchemaBuilder.FieldAssembler<Schema> builder =
						getNullableBuilder(logicalType)
								.record(rowName)
								.fields();
				for (int i = 0; i < rowType.getFieldCount(); i++) {
					String fieldName = fieldNames.get(i);
					builder = builder
						.name(fieldName)
						.type(convertToSchema(rowType.getTypeAt(i), rowName + "_" + fieldName))
						.noDefault();
				}
				return builder.endRecord();
```

File: flink-formats/flink-avro/src/test/java/org/apache/flink/formats/avro/typeutils/AvroSchemaConverterTest.java
Comment (dawidwys) [inline flink-formats/flink-avro/src/test/java/org/apache/flink/formats/avro/typeutils/AvroSchemaConverterTest.java:107]: I think it would be nice to add a test that we can convert back and forth between `DataType` and `Schema` in respect to the field names.

**Review by dawidwys — CHANGES_REQUESTED** [review]
Thanks for working on the issues @danny0405 .

You are fixing multiple (at least 2) independent issues in the PR. Could you please split it into separate commits with appropriate commit messages.

I also commented on the solution for records fields names.

File: flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroSchemaConverter.java
Comment (danny0405) [reply flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroSchemaConverter.java:436]: They are for different purpose but i think we can use `nullableSchema` altogether.

File: flink-formats/flink-avro/src/test/java/org/apache/flink/formats/avro/typeutils/AvroSchemaConverterTest.java
Comment (danny0405) [reply flink-formats/flink-avro/src/test/java/org/apache/flink/formats/avro/typeutils/AvroSchemaConverterTest.java:107]: Already added, see `AvroSchemaConverterTest.testConversionIntegralityNullable`.

**Comment by danny0405** [comment]
Thanks @dawidwys for the review, i have addressed your comments.

File: flink-formats/flink-avro/src/test/java/org/apache/flink/formats/avro/typeutils/AvroSchemaConverterTest.java
Comment (dawidwys) [reply flink-formats/flink-avro/src/test/java/org/apache/flink/formats/avro/typeutils/AvroSchemaConverterTest.java]: True, you test though only the `DataType` -> `Schema` and back. We could also add a test for `Schema` -> `DataType` (you start with a Schema), which is important in case of a schema registry.

File: flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroSchemaConverter.java
Comment (dawidwys) [inline flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroSchemaConverter.java:317]: Is it necessary? Shouldn't the `top` be handled in the planner and just passed with a correct setting in the `logicalType`? I think it is a bit too deep to fix it here.

**Review by dawidwys — CHANGES_REQUESTED** [review]
Thanks for the update @danny0405 ! I think it's in quite a good shape. I put some final comments. Could you also update the second commit with the right jira number. If I am correct it addresses FLINK-19786 instead of FLINK-19779

File: flink-formats/flink-avro/src/test/java/org/apache/flink/formats/avro/AvroRowDataDeSerializationSchemaTest.java
Comment (danny0405) [reply flink-formats/flink-avro/src/test/java/org/apache/flink/formats/avro/AvroRowDataDeSerializationSchemaTest.java:184]: Thanks, would remove it.

File: flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroSchemaConverter.java
Comment (danny0405) [reply flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroSchemaConverter.java:317]: It is meaningless to have a nullable top row type, and in the `CREATE TABLE` DDL there is no way to specify that the table row is not nullable, actually it should always be not null(even if all the fields are null).

File: flink-formats/flink-avro/src/test/java/org/apache/flink/formats/avro/typeutils/AvroSchemaConverterTest.java
Comment (danny0405) [reply flink-formats/flink-avro/src/test/java/org/apache/flink/formats/avro/typeutils/AvroSchemaConverterTest.java]: Thanks, i would add `schema` -> `DataType` -> `schema` test.

File: flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroSchemaConverter.java
Comment (dawidwys) [reply flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroSchemaConverter.java:317]: I agree that the top row should not be nullable. It is a property of a SQL Table though, not of the format. If I am not wrong there you can not specify a ROW nullable in `CREATE TABLE` because it is always `NOT NULL`. Therefore it is meaningless to have additional parameter in the `convertToSchema`.

What I am saying is that the format should not be bother with this. It is purely a matter of the planner.

Comment (wuchong) [reply flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroSchemaConverter.java:317]: I have the same feeling with @dawidwys .

Comment (danny0405) [reply flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroSchemaConverter.java:317]: Currently, only avro format cares about the outer row nullability, we can switch to change the planner if we found more user cases.

Technically to say, it is impossible to infer the outer row nullability only from the DDL.

Comment (dawidwys) [reply flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroSchemaConverter.java:317]: Sorry, but I don't understand your arguments.

What do you mean by "inferring" the outer row nullability? There is no inference. In SQL the outer row is always `NOT NULL` or otherwise the row simply does not exist.

In the current shape you have two arguments that contradict each other. Take this example:

```
DataType type = DataTypes.ROW(
    DataTypes.FIELD("f0", DataTypes.ROW(
        DataTypes.FIELD("f1", DataTypes.BIGINT()
    ).nullable()
).nullable();

Schema schema = convertToSchema(type, "record", true) // <- you're overriding the property of type
// you can achieve the same with
// Schema schema = convertToSchema(type.notNull(), "record")
```

It's the Planners task to produce a `notNull` type for a Tables type.

Comment (danny0405) [reply flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroSchemaConverter.java]: > `In SQL the outer row is always NOT NULL`
It is not true, a row (null, null) is nullable true. And i don't think it makes sense to change the planner behavior in general in order to fix a specific use case. 

>// you can achieve the same with
>// Schema schema = convertToSchema(type.notNull(), "record") 

I don't think we should let each invoker to decide whether to make the data type not null, because in current codebase, we should always do that, make the decision everyone is error-prone and hard to maintain.

I have merge the `top` into a row type nullability switch, see `AvroSchemaConverter.convertToSchema(LogicalType logicalType)`.

File: flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroSchemaConverter.java
Comment (wuchong) [inline flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroSchemaConverter.java:305]: I think we can make the parameter to be `RowType`, that would be make sense to use it as the top-level row type and not generate nullable for it. Besides, would be better to add comments in the Javadoc. Currently, this method has the same Javadoc with `convertToSchema(LogicalType logicalType, String rowName)`.

Comment (danny0405) [reply flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroSchemaConverter.java:308]: Although the `AvroSchemaConverter` is a tool class, it is still used as a public API, so i'm inclined to keep the signature unchanged. Another reason is that only logical type is enough for the conversion.

Have added more documents to the methods.

File: flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroSchemaConverter.java
Comment (dawidwys) [reply flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroSchemaConverter.java]: > It is not true, a row (null, null) is nullable true. And i don't think it makes sense to change the planner behavior in general in order to fix a specific use case.

Excuse me, but I wholeheartedly disagree with your statement. null =/= (null, null). (null, null) is still `NOT NULL`. A whole row in a Table can not be null. Only particular columns can be null. Therefore the top level row of a Table is always `NOT NULL`.

I am not suggesting changing planner behaviour for a particular use case. The planner should always produce `NOT NULL` type for a top level row of a Table. If it doesn't, it is a bug.

Comment (dawidwys) [reply flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroSchemaConverter.java]: > I don't think we should let each invoker to decide whether to make the data type not null, because in current codebase, we should always do that, make the decision everyone is error-prone and hard to maintain.

I agree making the same decision over and over again at multiple location is error prone and hard to maintain and that's what I want to avoid.

Comment (danny0405) [reply flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroSchemaConverter.java]: >Excuse me, but I wholeheartedly disagree with your statement. null =/= (null, null). (null, null) is still NOT NULL. A whole row in a Table can not be null. Only particular columns can be null. Therefore the top level row of a Table is always NOT NULL.

You can test it in PostgreSQL with the following SQL:
```sql
create type my_type as (a int, b varchar(20));

create table t1(
  f0 my_type,
  f1 varchar(20)
);

insert into t1 values((null, null), 'def');

select f0 is null from t1; -- it returns true
```

Comment (dawidwys) [reply flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroSchemaConverter.java]: But the `my_type` is not a top level row in your example. It is a type of a column. It has nothing to do with the case we're discussing.

Comment (danny0405) [reply flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroSchemaConverter.java]: I had offline discussion with @wuchong and @dawidwys , and after some research, we found that a non-nullable row type is more reasonable.

But because the change is huge(many codes that convert a type info to data type assumes nullable true before), me and @wuchong decide to change the method signature to `convertToSchema(RowType schema)` and add a notion to the method doc that the passed in `schema` must be the top level record type.

Comment (danny0405) [reply flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroSchemaConverter.java]: @dawidwys  I have changed the schema row type to be always nullable false, please take a look again if you have time, thanks so much.

**Comment by dawidwys** [comment]
I think there is still a test failling: `org.apache.flink.table.client.gateway.local.LocalExecutorITCase#testStreamQueryExecutionSink`

**Comment by danny0405** [comment]
> I think there is still a test failling: `org.apache.flink.table.client.gateway.local.LocalExecutorITCase#testStreamQueryExecutionSink`

Yes, sorry for that i forget to commit some files, re-commit and run the testing ~

**Review by dawidwys — APPROVED** [review]
+1, Thanks for the update @danny0405 I'll merge once azure gives green.

**Pull request merged**
